source/common/network/transport_socket_options_impl.cc

```diff
@@ -43,9 +46,16 @@ TransportSocketOptionsUtility::fromFilterState(const StreamInfo::FilterState& fi
     needs_transport_socket_options = true;
   }
 
+  if (filter_state.hasData<UpstreamSubjectAltNames>(UpstreamSubjectAltNames::key())) {
+    const auto& upstream_subject_alt_names =
+        filter_state.getDataReadOnly<UpstreamSubjectAltNames>(UpstreamSubjectAltNames::key());
+    subject_alt_names = upstream_subject_alt_names.value();
+    needs_transport_socket_options = true;
+  }
+
   if (needs_transport_socket_options) {
     return std::make_shared<Network::TransportSocketOptionsImpl>(
-        server_name, std::vector<std::string>{}, std::move(application_protocols));
+        server_name, std::move(subject_alt_names), std::move(application_protocols));
```

Comment:
Hmm, I think `TransportSocketOptionsImpl::hashKey` above needs to get modified to take into account `override_verify_san_list_`, otherwise we won't necessarily correctly split the connection if for some reason SNI != the host to verify? I think this is pedantic for the cases we have now where they are always equal but you might as well fix and test that case?

Reply:
I will do this as a separate PR, though to be clear this isn't an issue right now even with this PR ( (except out-of-tree filters setting UpstreamSubjectAltNames individually), because we won't have two connections created in same cluster have same SNI but with different SAN list, it could be SNI != host to verify while it won't vary.

Reply:
Agreed it's pedantic for this PR. Can you add a TODO though above?